### ISSUE TITLE
enhancement : tar : .tar files as a filesystem tree replacement

### ISSUE THREAD

**Issue by kmeaw** [folbricht/desync#84]
There are various programs (such as OS generators) which are capable of generating .tar format streams instead of populating the filesystem tree.
It would be nice if a user could pipe such program into "desync tar" directly instead of unpacking the stream into a temporary storage and then running desync on the top of it.

**Comment by folbricht** [comment]
@kmeaw With https://github.com/folbricht/desync/pull/85 merged, is this still useful?

**Comment by kmeaw** [comment]
#85 allows the user to install desync directly into the target filesystem tree and then run `desync tar caidx / -x`. This feature would not require desync to be inside the tree to be packed.
An example of the source of tar stream is a Docker Registry.

**Comment by charles-dyfis-net** [comment]
@kmeaw, by "tar", you mean GNU tar / us-tar / etc?

caidx format has a number of extensions to allow efficient random-access lookup. tar, by contrast, was designed for tape drives, where random access wasn't available -- instead, you were always going to stream from one end to the other.

I don't see a reasonable way to make desync support tar archives.

**Comment by kmeaw** [comment]
Yes, I mean us-tar file format.

tar is used as a serialization format for an image layer in Docker. To import a catar+caidx into Docker, I would need to create a temporary directory, run desync to unpack, then run tar to serialize and pipe it into "docker import -".

**Comment by folbricht** [comment]
Perhaps I misunderstood the request previously. So the goal is to read from/write to GNU tar files instead of the filesystem when chunking/extracting a catar? Somthing like this:

Read files from a GNU tar file, chunk them and turn into a catar/caidx:
```
desync tar -i file.caidx --input-format=tar file.tar

desync tar -i file.caidx --input-format=tar -
```

And extracting:

```
desync untar -i file.caidx --ouput-format=tar file.tar

desync untar -i file.caidx --output-format=tar -
```

**Comment by folbricht** [comment]
Started working on this on https://github.com/folbricht/desync/tree/issue-84 with abstracting the output filesystem since that seems a little easier. I might need someone to help/review the xattr bits later.

**Comment by folbricht** [comment]
The untar to GNU tar part of this has been implemented on the branch. Can be used like so:

```
desync untar --output-format=gnu-tar input.catar output.tar
```

Also works with caidx indexes (-i option) as before. The target can be - to write to STDOUT. Note that this also fixes a bug in the device-creation during untar which was introduced last year.

Any feedback is much appreciated.

**Comment by tomberek** [comment]
desync untar --output-format=gnu-tar works on .catar
also tested on .caidx w/ a store, it works!

**Comment by folbricht** [comment]
Thanks for trying that out. The next step would be for me to add reading from tar archives and produce catar/caidx

**Comment by tomberek** [comment]
Ya, something along the lines of tar2sqfs (https://github.com/AgentD/squashfs-tools-ng) would make it easy to adopt. Then there is still #12, mounting. I've tried to use desync for most operations, then casync for it's mount verb, with mixed success. I think there is an incompatibility somewhere, but i don't have a minimal case yet.

**Comment by folbricht** [comment]
A first version of reading/writing tar archives can be found in https://github.com/folbricht/desync/pull/118

**Comment by folbricht** [comment]
@kmeaw Any chance you could try out the implementation before it I merge it? Just to be sure it meets the use-case you described above.

**Comment by folbricht** [comment]
Implemented in #118
